presto-hive/src/main/java/io/prestosql/plugin/hive/proxy/LoggingInvocationHandler.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive.proxy;
+
+import com.google.common.reflect.AbstractInvocationHandler;
+import io.airlift.units.Duration;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import java.util.Arrays;
+import java.util.function.Consumer;
+import java.util.regex.Pattern;
+import java.util.stream.IntStream;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+
+public class LoggingInvocationHandler
+        extends AbstractInvocationHandler
+{
+    private static final Pattern TRAILING_DIGITS = Pattern.compile("\\d+$");
+
+    private final Object delegate;
+    private final Consumer<String> logger;
+
+    public LoggingInvocationHandler(Object delegate, Consumer<String> logger)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        this.logger = requireNonNull(logger, "logger is null");
+    }
+
+    @Override
+    protected Object handleInvocation(Object proxy, Method method, Object[] args)
+            throws Throwable
+    {
+        Object result;
+        long startNanos = System.nanoTime();
+        try {
+            result = method.invoke(delegate, args);
+        }
+        catch (Exception e) {
+            Duration elapsed = Duration.nanosSince(startNanos);
+            logger.accept(format("%s took %s and failed with %s", invocationDescription(method, args), elapsed, e));
+            throw e;
+        }
+        Duration elapsed = Duration.nanosSince(startNanos);
+        logger.accept(format("%s succeeded in %s", invocationDescription(method, args), elapsed));
```

Comment:
You should not format messages that are not going to be logged.

Instead of using `Consumer<String> logger` you could have a dedicated consumer interface with two methods for successful or not successful method call. So exception could be passed as first argument to `log.debug(e,...)`

Reply:
> You should not format messages that are not going to be logged.

This whole mechanism is used only when messages are going to be logged.

> you could have a dedicated consumer interface with two methods for successful or not successful method call. So exception could be passed as first argument to log.debug(e,...)

Is there a point in logging full exception when the call fails?
The call will be retried. If it fails, all exceptions will be logged.